src/Syntax/Types.hs

```diff
@@ -40,119 +45,80 @@ instance Show (SType a) where
     SBoolean -> "bool"
     SByteStr -> "bytestring"
 
-instance TestEquality SType where
-  testEquality SType SType = eqT
+type instance Sing = SType
 
--- | Compare equality of two things parametrized by types which have singletons.
-eqS :: forall (a :: *) (b :: *) f t. (SingI a, SingI b, Eq (f a t)) => f a t -> f b t -> Bool
-eqS fa fb = maybe False (\Refl -> fa == fb) $ testEquality (sing @a) (sing @b)
+instance TestEquality SType where
+  testEquality SInteger SInteger = Just Refl
+  testEquality SBoolean SBoolean = Just Refl
+  testEquality SByteStr SByteStr = Just Refl
+  testEquality _ _ = Nothing
 
--- | For our purposes, the singleton of a type @a@ is always @'SType' a@.
--- We need this to be able to use 'SomeSing' when implementing 'ActType'.
 
--- Note that even though there only exist three different 'SType', this does
--- not mean that the type family is partial. It simply means that the resulting
--- type is uninhabited if the argument is neither 'Integer', 'Bool' nor
--- 'ByteString'.
-type instance Sing = SType
+-- -- | Compare equality of two things parametrized by types which have singletons.
+-- eqS :: forall (a :: *) (b :: *) f t. (SingI a, SingI b, Eq (f a t)) => f a t -> f b t -> Bool
+-- eqS fa fb = maybe False (\Refl -> fa == fb) $ testEquality (sing @a) (sing @b)
```

Comment:
```suggestion
```

Is this needed?

Reply:
It was the old, and more concise definition of `eqS`. I'm bringing this back, thanks for noticing.